### PR TITLE
Simple function call support

### PR DESCRIPTION
**Pull request by jacob-hughes**
In this branch, I've tidied up the test suite a bit. The output was getting ridiculous and it was hard to see the wood for the trees. In this version, if a test rust program fails, stderr is captured and displayed at the end of the test runner.

I've also included functionality for basic function call support.

### CODE REVIEW

File: tests/runtests.rs [hunk 1]
```diff
@@ -68,26 +54,69 @@ fn get_sysroot() -> PathBuf {
...
-                    .expect("mirkat bin not found");
-    status.code().unwrap()
+/// Invokes Mirkat when given a path to a valid Rust program. Compares the
+/// expected return code against that returned from Mirkat. This is surprisingly
+/// effective at catching bugs in Mirkat. It's not uncommon that when things go
+/// wrong, Mirkat blows up catastrophically.
+fn test_prog(path: &Path, expected_code: i32) -> Result<(), Vec<u8>> {
+    let out = run_prog(path);
+
+    match out.status.code().unwrap() {
```
Comment (ltratt) [inline tests/runtests.rs]: Wouldn't this be clearer as a traditional `if`?

**Review by ltratt — COMMENTED** [review]
Pooped "in" reverse order?

File: src/machine.rs [hunk 2]
```diff
@@ -196,14 +203,30 @@ impl<'a, 'tcx> Machine<'a, 'tcx> {
...
                       def_id: DefId,
-                      ret_addr: Option<BasicBlock>) {
+                      mut args: Vec<TypedVal<'tcx>>,
+                      ret_val: Option<Address>,
+                      ret_block: Option<BasicBlock>){
         let def = Instance::mono(self.tcx, def_id).def;
         let mir = self.load_mir(def);
-        let frame = Frame::new(def_id, mir, ret_addr);
+
+
```
Comment (ltratt) [inline src/machine.rs]: Double blank line.

File: src/machine.rs [hunk 3]
```diff
@@ -196,14 +203,30 @@ impl<'a, 'tcx> Machine<'a, 'tcx> {
...
+        // Initialise locals
+        let size = mir.local_decls.len();
+        let mut tmp_locals = vec![None; size]; // Init to none
```
Comment (ltratt) [inline src/machine.rs]: I think the comment is obvious enough that it's not useful.

Comment (jacob-hughes) [reply src/machine.rs]: True. Removed

File: src/machine.rs (hunk 2 above)
Comment (jacob-hughes) [reply src/machine.rs]: Fixed

File: tests/runtests.rs (hunk 1 above)
Comment (jacob-hughes) [reply tests/runtests.rs]: Yes it would, I did adjust this later to fix this. When I squash I will merge these into a single commit

**Comment by ltratt** [comment]
Looks fine to me other than those minor comments.

**Comment by jacob-hughes** [comment]
Thanks. W.r.t your comment about popping args in reverse order (With a fantastic typo I might add ;) ), I agree this was poorly worded and I'll fix up this for clarity when squashing down. I've made suggested fixes, let me know if you're happy for me to squash.

**Comment by ltratt** [comment]
I think we need to wait for @vext01's comments first.

**Comment by jacob-hughes** [comment]
Ah yes jumped the gun a bit there. No problem

File: src/interp.rs [hunk 4]
```diff
@@ -145,15 +150,46 @@ impl<'a, 'tcx> Machine<'a, 'tcx> {
...
+                let (fn_def, sig) = match func.ty.sty {
+                    ty::TyFnDef(def_id, substs) => {
+                        let instance = ty::Instance::resolve(
+                            self.tcx,
+                            self.tcx.param_env(def_id),
+                            def_id,
+                            substs,
+                        ).unwrap();
+                        let sig = func.ty.fn_sig(self.tcx);
+                        (instance, sig)
```
Comment (vext01) [inline src/interp.rs]: Is `sig` used?

File: src/machine.rs [hunk 5]
```diff
@@ -66,17 +72,17 @@ pub struct Frame<'tcx> {
...
-               ret_addr: Option<BasicBlock>)
+               locals: IndexVec<Local,Option<Vec<u8>>>,
+               ret_val: Option<Address>,
+               ret_block: Option<BasicBlock>)
                -> Frame<'tcx> {
-        let size = mir.local_decls.len();
-        let mut locals:IndexVec<Local, Option<Vec<u8>>> = IndexVec::with_capacity(size);
-        locals.extend(vec![None; size]);
 
         Frame {
```
Comment (vext01) [inline src/machine.rs]: For future reference, you can use `Self { ... }` which makes the code easier to maintain (e.g. easier to rename structs later).

**Review by vext01 — COMMENTED** [review]
This looks good to me. A couple of small comments.

File: src/machine.rs (hunk 5 above)
Comment (jacob-hughes) [reply src/machine.rs]: That's neat. I didn't know this!

File: src/interp.rs (hunk 4 above)
Comment (jacob-hughes) [reply src/interp.rs]: Nope, good spot, it's leftover code from a previous approach to implementation!

**Comment by jacob-hughes** [comment]
Changes made, can I squash?

**Comment by vext01** [comment]
LGTM. Go ahead.

**Comment by jacob-hughes** [comment]
Squashed.

**Pull request merged**